File: javax/src/test/java/me/panpf/javax/test/lang/ClassxTest.java
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.javax.test.lang;

import me.panpf.javax.lang.Classx;
import me.panpf.javax.util.Arrayx;
import me.panpf.javax.util.Predicate;
import org.jetbrains.annotations.NotNull;
import org.junit.Assert;
import org.junit.Test;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("unused")
public class ClassxTest {

    public static class TestField1 {
        public String testFiled11 = "field11";
        public String testFiled12 = "field12";
    }

    public static class TestField2 extends TestField1 {
        public String testFiled21 = "field21";
        public String testFiled22 = "field22";
    }

    public static class TestField3 extends TestField2 {
        public String testFiled31 = "field31";
        public String testFiled32 = "field32";
    }

    @SuppressWarnings("WeakerAccess")
    public static class TestStatic {
        public static String filed1 = "filed11";

        public static String test() {
            return filed1;
        }
    }

    public static class TestMethod {

        private String update;

        public String getUpdate() {
            return update;
        }

        public void setUpdate(String update) {
            this.update = update;
        }
    }

    @SuppressWarnings("WeakerAccess")
    public static class TestConstructor {

        public TestConstructor() {
        }

        public TestConstructor(String param1) {
        }

        public TestConstructor(String param1, int param2) {
        }
    }

    public static class TestType {
        String[] strings = new String[]{};
        List<String> stringList = new ArrayList<>();
    }

    @Test
    public void testField() throws NoSuchFieldException, ClassNotFoundException {
        /*
         * getFieldWithParent
         */
        Assert.assertNotNull(Classx.getFieldWithParent(TestStatic.class, "filed1"));
        try {
            Classx.getFieldWithParent(TestStatic.class, "filed_no");
            Assert.fail();
        } catch (Exception ignored) {
        }

        Assert.assertNotNull(Classx.getFieldWithParent(new TestStatic(), "filed1"));
        try {
            Classx.getFieldWithParent(new TestStatic(), "filed_no");
            Assert.fail();
        } catch (Exception ignored) {
        }

        Assert.assertNotNull(Classx.getFieldWithParent("me.panpf.javax.test.lang.ClassxTest$TestStatic", "filed1"));
        try {
            Classx.getFieldWithParent("me.panpf.javax.test.lang.ClassxTest$TestStatic", "filed_no");
            Assert.fail();
        } catch (Exception ignored) {
        }

        /*
         * getFieldsWithParent
         */
        Predicate<Field> filterJacocoFiledPredicate = new Predicate<Field>() {
            @Override
            public boolean accept(@NotNull Field field) {
                return !field.getName().equals("$jacocoData");
            }
        };

        Assert.assertEquals(Arrayx.filter(Classx.getFieldsWithParent(TestField3.class, 1), filterJacocoFiledPredicate).toArray(new Field[0]).length, 4);
        Assert.assertEquals(Arrayx.filter(Classx.getFieldsWithParent(TestField3.class), filterJacocoFiledPredicate).toArray(new Field[0]).length, 6);

        Assert.assertEquals(Arrayx.filter(Classx.getFieldsWithParent(new TestField3(), 1), filterJacocoFiledPredicate).toArray(new Field[0]).length, 4);
        Assert.assertEquals(Arrayx.filter(Classx.getFieldsWithParent(new TestField3()), filterJacocoFiledPredicate).toArray(new Field[0]).length, 6);

        Assert.assertEquals(Arrayx.filter(Classx.getFieldsWithParent("me.panpf.javax.test.lang.ClassxTest$TestField3", 1), filterJacocoFiledPredicate).toArray(new Field[0]).length, 4);
        Assert.assertEquals(Arrayx.filter(Classx.getFieldsWithParent("me.panpf.javax.test.lang.ClassxTest$TestField3"), filterJacocoFiledPredicate).toArray(new Field[0]).length, 6);


        /*
         * get and set field value
         */

        TestField1 testField1 = new TestField1();

        Classx.setFieldValue(testField1, Classx.getFieldWithParent(testField1, "testFiled11"), "field11x");
        Assert.assertEquals(Classx.getFieldValue(testField1, Classx.getFieldWithParent(testField1, "testFiled11")), "field11x");

        Classx.setFieldValue(testField1, "testFiled11", "field11y");
        Assert.assertEquals(Classx.getFieldValue(testField1, "testFiled11"), "field11y");


        /*
         * get and set static filed value
         */

        Classx.setStaticFieldValue(Classx.getFieldWithParent(TestStatic.class, "filed1"), "filed12");
        Assert.assertEquals(Classx.getStaticFieldValue(Classx.getFieldWithParent(TestStatic.class, "filed1")), "filed12");

        Classx.setStaticFieldValue(TestStatic.class, "filed1", "filed13");
        Assert.assertEquals(Classx.getStaticFieldValue(TestStatic.class, "filed1"), "filed13");

        Classx.setStaticFieldValue("me.panpf.javax.test.lang.ClassxTest$TestStatic", "filed1", "filed15");
        Assert.assertEquals(Classx.getStaticFieldValue("me.panpf.javax.test.lang.ClassxTest$TestStatic", "filed1"), "filed15");
    }

    @Test
    public void testMethod() throws NoSuchMethodException, ClassNotFoundException, NoSuchFieldException {
        /*
         * getMethodWithParent
         */

        Assert.assertNotNull(Classx.getMethodWithParent(TestStatic.class, "test"));
        try {
            Classx.getMethodWithParent(TestStatic.class, "test_no");
            Assert.fail();
        } catch (NoSuchMethodException ignored) {
        }

        Assert.assertNotNull(Classx.getMethodWithParent(new TestStatic(), "test"));
        try {
            Classx.getMethodWithParent(new TestStatic(), "test_no");
            Assert.fail();
        } catch (NoSuchMethodException ignored) {
        }

        Assert.assertNotNull(Classx.getMethodWithParent("me.panpf.javax.test.lang.ClassxTest$TestStatic", "test"));
        try {
            Classx.getMethodWithParent("me.panpf.javax.test.lang.ClassxTest$TestStatic", "test_no");
            Assert.fail();
        } catch (NoSuchMethodException ignored) {
        }

        /*
         * getMethodsWithParent
         */

        Predicate<Method> filterJacocoMethodPredicate = new Predicate<Method>() {
            @Override
            public boolean accept(@NotNull Method field) {
                return !field.getName().equals("$jacocoInit");
            }
        };

        Assert.assertEquals(Arrayx.filter(Classx.getMethodsWithParent(TestMethod.class), filterJacocoMethodPredicate).toArray(new Method[0]).length, 14);
        Assert.assertEquals(Arrayx.filter(Classx.getMethodsWithParent(TestMethod.class, 0), filterJacocoMethodPredicate).toArray(new Method[0]).length, 2);

        Assert.assertEquals(Arrayx.filter(Classx.getMethodsWithParent(new TestMethod()), filterJacocoMethodPredicate).toArray(new Method[0]).length, 14);
        Assert.assertEquals(Arrayx.filter(Classx.getMethodsWithParent(new TestMethod(), 0), filterJacocoMethodPredicate).toArray(new Method[0]).length, 2);

        Assert.assertEquals(Arrayx.filter(Classx.getMethodsWithParent("me.panpf.javax.test.lang.ClassxTest$TestMethod"), filterJacocoMethodPredicate).toArray(new Method[0]).length, 14);
        Assert.assertEquals(Arrayx.filter(Classx.getMethodsWithParent("me.panpf.javax.test.lang.ClassxTest$TestMethod", 0), filterJacocoMethodPredicate).toArray(new Method[0]).length, 2);

        /*
         * callMethod
         */

        TestMethod testMethod = new TestMethod();

        Classx.callMethod(testMethod, Classx.getMethodWithParent(testMethod, "setUpdate", String.class), "updatex");
        Assert.assertEquals(Classx.callMethod(testMethod, Classx.getMethodWithParent(testMethod, "getUpdate")), "updatex");

        Classx.callMethod(testMethod, "setUpdate", "updatex");
        Assert.assertEquals(Classx.callMethod(testMethod, "getUpdate"), "updatex");


        /*
         * callStaticMethod
         */
        Classx.setStaticFieldValue(TestStatic.class, "filed1", "filedx");

        Assert.assertEquals(Classx.callStaticMethod(Classx.getMethodWithParent(TestStatic.class, "test")), "filedx");
        Assert.assertEquals(Classx.callStaticMethod(TestStatic.class, "test"), "filedx");
        Assert.assertEquals(Classx.callStaticMethod("me.panpf.javax.test.lang.ClassxTest$TestStatic", "test"), "filedx");
    }

    @Test
    public void testConstructor() throws NoSuchMethodException, ClassNotFoundException {
        Assert.assertNotNull(Classx.getConstructorWithParent(TestConstructor.class, String.class));
        Assert.assertNotNull(Classx.getConstructorWithParent(new TestConstructor(), String.class));
        Assert.assertNotNull(Classx.getConstructorWithParent("me.panpf.javax.test.lang.ClassxTest$TestConstructor", String.class));

        Assert.assertEquals(Classx.getConstructorsWithParent(TestConstructor.class).length, 4);
        Assert.assertEquals(Classx.getConstructorsWithParent(TestConstructor.class, 0).length, 3);

        Assert.assertEquals(Classx.getConstructorsWithParent(new TestConstructor()).length, 4);
        Assert.assertEquals(Classx.getConstructorsWithParent(new TestConstructor(), 0).length, 3);

        Assert.assertEquals(Classx.getConstructorsWithParent("me.panpf.javax.test.lang.ClassxTest$TestConstructor").length, 4);
        Assert.assertEquals(Classx.getConstructorsWithParent("me.panpf.javax.test.lang.ClassxTest$TestConstructor", 0).length, 3);
    }

    @Test
    public void testHierarchy() throws ClassNotFoundException {
        Assert.assertEquals(Classx.getClassHierarchy(TestField3.class).length, 4);
        Assert.assertEquals(Classx.getClassHierarchy(TestField3.class, true).length, 3);

        Assert.assertEquals(Classx.getClassHierarchy(new TestField3()).length, 4);
        Assert.assertEquals(Classx.getClassHierarchy(new TestField3(), true).length, 3);

        Assert.assertEquals(Classx.getClassHierarchy("me.panpf.javax.test.lang.ClassxTest$TestField3").length, 4);
        Assert.assertEquals(Classx.getClassHierarchy("me.panpf.javax.test.lang.ClassxTest$TestField3", true).length, 3);
    }

    @Test
    public void testType() {
        try {
            Assert.assertTrue(Classx.isTypeArray(Classx.getFieldWithParent(TestType.class, "strings"), String.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertFalse(Classx.isTypeArray(Classx.getFieldWithParent(TestType.class, "strings"), Integer.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertTrue(Classx.isTypeCollection(Classx.getFieldWithParent(TestType.class, "stringList"), List.class, String.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertFalse(Classx.isTypeCollection(Classx.getFieldWithParent(TestType.class, "stringList"), List.class, Integer.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: javax-kt/src/test/java/me/panpf/javaxkt/test/lang/ClassxTest.kt
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.javaxkt.test.lang

import me.panpf.javax.collections.Arrayx
import me.panpf.javax.util.Predicate
import me.panpf.javaxkt.lang.*
import org.junit.Assert
import org.junit.Test
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.util.*

class ClassxTest {

    open class TestField1 {
        var testField11 = "field11"
        var testField12 = "field12"
    }

    @Suppress("unused")
    open class TestField2 : TestField1() {
        var testField21 = "field21"
        var testField22 = "field22"
    }

    @Suppress("unused")
    class TestField3 : TestField2() {
        var testField31 = "field31"
        var testField32 = "field32"
    }

    @Suppress("unused")
    class TestMethod {
        var update: String? = null
    }

    @Suppress("unused")
    class TestStatic {
        companion object {
            var field1 = "field11"

            @JvmStatic
            fun test() = field1
        }
    }

    @Suppress("unused")
    class TestConstructor {
        constructor()

        @Suppress("UNUSED_PARAMETER")
        constructor(param1: String)

        @Suppress("UNUSED_PARAMETER")
        constructor(param1: String, param2: Int)
    }

    @Suppress("unused")
    class TestType {
        internal var strings = arrayOf<String>()
        internal var stringList: List<String> = ArrayList()
    }

    @Test
    @Throws(NoSuchFieldException::class, ClassNotFoundException::class)
    fun testField() {
        /*
         * getFieldWithParent
         */
        Assert.assertNotNull(TestStatic::class.java.getFieldWithParent("field1"))
        try {
            TestStatic::class.java.getFieldWithParent("field_no")
            Assert.fail()
        } catch (ignored: Exception) {
        }

        Assert.assertNotNull(TestStatic().getFieldWithParent("field1"))
        try {
            TestStatic().getFieldWithParent("field_no")
            Assert.fail()
        } catch (ignored: Exception) {
        }

        Assert.assertNotNull("me.panpf.javaxkt.test.lang.ClassxTest\$TestStatic".getFieldWithParent("field1"))
        try {
            "me.panpf.javaxkt.test.lang.ClassxTest\$TestStatic".getFieldWithParent("field_no")
            Assert.fail()
        } catch (ignored: Exception) {
        }

        /*
         * getFieldsWithParent
         */
        val filterJacocoFieldPredicate = Predicate<Field> { field -> field.name != "\$jacocoData" }

        Assert.assertEquals(Arrayx.filter(TestField3::class.java.getFieldsWithParent(1), filterJacocoFieldPredicate).toTypedArray().size.toLong(), 4)
        Assert.assertEquals(Arrayx.filter(TestField3::class.java.getFieldsWithParent(), filterJacocoFieldPredicate).toTypedArray().size.toLong(), 6)

        Assert.assertEquals(Arrayx.filter(TestField3().getFieldsWithParent(1), filterJacocoFieldPredicate).toTypedArray().size.toLong(), 4)
        Assert.assertEquals(Arrayx.filter(TestField3().getFieldsWithParent(), filterJacocoFieldPredicate).toTypedArray().size.toLong(), 6)

        Assert.assertEquals(Arrayx.filter("me.panpf.javaxkt.test.lang.ClassxTest\$TestField3".getFieldsWithParent(1), filterJacocoFieldPredicate).toTypedArray().size.toLong(), 4)
        Assert.assertEquals(Arrayx.filter("me.panpf.javaxkt.test.lang.ClassxTest\$TestField3".getFieldsWithParent(), filterJacocoFieldPredicate).toTypedArray().size.toLong(), 6)


        /*
         * get and set field value
         */

        val testField1 = TestField1()

        testField1.setFieldValue(testField1.getFieldWithParent("testField11"), "field11x")
        Assert.assertEquals(testField1.getFieldValue(testField1.getFieldWithParent("testField11")), "field11x")

        testField1.setFieldValue("testField11", "field11y")
        Assert.assertEquals(testField1.getFieldValue("testField11"), "field11y")


        /*
         * get and set static field value
         */

        TestStatic::class.java.getFieldWithParent("field1").setStaticValue("field12")
        Assert.assertEquals(TestStatic::class.java.getFieldWithParent("field1").getStaticValue(), "field12")

        TestStatic::class.java.setStaticFieldValue("field1", "field13")
        Assert.assertEquals(TestStatic::class.java.getStaticFieldValue("field1"), "field13")

        "me.panpf.javaxkt.test.lang.ClassxTest\$TestStatic".setStaticFieldValue("field1", "field15")
        Assert.assertEquals("me.panpf.javaxkt.test.lang.ClassxTest\$TestStatic".getStaticFieldValue("field1"), "field15")
    }

    @Test
    @Throws(NoSuchMethodException::class, ClassNotFoundException::class, NoSuchFieldException::class)
    fun testMethod() {
        /*
         * getMethodWithParent
         */

        Assert.assertNotNull(TestStatic::class.java.getMethodWithParent("test"))
        try {
            TestStatic::class.java.getMethodWithParent("test_no")
            Assert.fail()
        } catch (ignored: NoSuchMethodException) {
        }

        Assert.assertNotNull(TestStatic().getMethodWithParent("test"))
        try {
            TestStatic().getMethodWithParent("test_no")
            Assert.fail()
        } catch (ignored: NoSuchMethodException) {
        }

        Assert.assertNotNull("me.panpf.javaxkt.test.lang.ClassxTest\$TestStatic".getMethodWithParent("test"))
        try {
            "me.panpf.javaxkt.test.lang.ClassxTest\$TestStatic".getMethodWithParent("test_no")
            Assert.fail()
        } catch (ignored: NoSuchMethodException) {
        }

        /*
         * getMethodsWithParent
         */

        val filterJacocoMethodPredicate = Predicate<Method> { field -> field.name != "\$jacocoInit" }

        Assert.assertEquals(Arrayx.filter(TestMethod::class.java.getMethodsWithParent(), filterJacocoMethodPredicate).toTypedArray().size.toLong(), 14)
        Assert.assertEquals(Arrayx.filter(TestMethod::class.java.getMethodsWithParent(0), filterJacocoMethodPredicate).toTypedArray().size.toLong(), 2)

        Assert.assertEquals(Arrayx.filter(TestMethod().getMethodsWithParent(), filterJacocoMethodPredicate).toTypedArray().size.toLong(), 14)
        Assert.assertEquals(Arrayx.filter(TestMethod().getMethodsWithParent(0), filterJacocoMethodPredicate).toTypedArray().size.toLong(), 2)

        Assert.assertEquals(Arrayx.filter("me.panpf.javaxkt.test.lang.ClassxTest\$TestMethod".getMethodsWithParent(), filterJacocoMethodPredicate).toTypedArray().size.toLong(), 14)
        Assert.assertEquals(Arrayx.filter("me.panpf.javaxkt.test.lang.ClassxTest\$TestMethod".getMethodsWithParent(0), filterJacocoMethodPredicate).toTypedArray().size.toLong(), 2)

        /*
         * callMethod
         */

        val testMethod = TestMethod()

        testMethod.callMethod(testMethod.getMethodWithParent("setUpdate", String::class.java), "updatex")
        Assert.assertEquals(testMethod.callMethod(testMethod.getMethodWithParent("getUpdate")), "updatex")

        testMethod.callMethod("setUpdate", "updatex")
        Assert.assertEquals(testMethod.callMethod("getUpdate"), "updatex")


        /*
         * callStaticMethod
         */
        TestStatic::class.java.setStaticFieldValue("field1", "fieldx")

        Assert.assertEquals(TestStatic::class.java.getMethodWithParent("test").callStaticMethod(), "fieldx")
        Assert.assertEquals(TestStatic::class.java.callStaticMethod("test"), "fieldx")
        Assert.assertEquals("me.panpf.javaxkt.test.lang.ClassxTest\$TestStatic".callStaticMethod("test"), "fieldx")
    }

    @Test
    @Throws(NoSuchMethodException::class, ClassNotFoundException::class)
    fun testConstructor() {
        Assert.assertNotNull(TestConstructor::class.java.getConstructorWithParent(String::class.java))
        Assert.assertNotNull(TestConstructor().getConstructorWithParent(String::class.java))
        Assert.assertNotNull("me.panpf.javaxkt.test.lang.ClassxTest\$TestConstructor".getConstructorWithParent(String::class.java))

        Assert.assertEquals(TestConstructor::class.java.getConstructorsWithParent().size.toLong(), 4)
        Assert.assertEquals(TestConstructor::class.java.getConstructorsWithParent(0).size.toLong(), 3)

        Assert.assertEquals(TestConstructor().getConstructorsWithParent().size.toLong(), 4)
        Assert.assertEquals(TestConstructor().getConstructorsWithParent(0).size.toLong(), 3)

        Assert.assertEquals("me.panpf.javaxkt.test.lang.ClassxTest\$TestConstructor".getConstructorsWithParent().size.toLong(), 4)
        Assert.assertEquals("me.panpf.javaxkt.test.lang.ClassxTest\$TestConstructor".getConstructorsWithParent(0).size.toLong(), 3)
    }

    @Test
    @Throws(ClassNotFoundException::class)
    fun testHierarchy() {
        Assert.assertEquals(TestField3::class.java.getClassHierarchy().size.toLong(), 4)
        Assert.assertEquals(TestField3::class.java.getClassHierarchy(true).size.toLong(), 3)

        Assert.assertEquals(TestField3().getClassHierarchy().size.toLong(), 4)
        Assert.assertEquals(TestField3().getClassHierarchy(true).size.toLong(), 3)

        Assert.assertEquals("me.panpf.javaxkt.test.lang.ClassxTest\$TestField3".getClassHierarchy().size.toLong(), 4)
        Assert.assertEquals("me.panpf.javaxkt.test.lang.ClassxTest\$TestField3".getClassHierarchy(true).size.toLong(), 3)
    }

    @Test
    fun testType() {
        try {
            Assert.assertTrue(TestType::class.java.getFieldWithParent("strings").isTypeArray(String::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertFalse(TestType::class.java.getFieldWithParent("strings").isTypeArray(Int::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertTrue(TestType::class.java.getFieldWithParent("stringList").isTypeCollection(List::class.java, String::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertFalse(TestType::class.java.getFieldWithParent("stringList").isTypeCollection(List::class.java, Int::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

    }
}
